packages/@lwc/engine-core/src/framework/modules/static-parts.ts

```diff
@@ -105,3 +104,39 @@ export function applyStaticParts(
         applyRefs(part, owner);
     }
 }
+
+/**
+ * Reapplies static parts mounting to newly generated vnodes.
+ *
+ * @param n1 - the previous VStatic vnode
+ * @param n2 - the current VStatic vnode
+ * @param root - the root element
+ * @param renderer - the renderer to use
+ */
+export function patchStaticParts(n1: VStatic, n2: VStatic, root: Element, renderer: RendererAPI) {
+    if (!process.env.IS_BROWSER) {
+        return;
+    }
+
+    const { parts: currParts, owner: currPartsOwner } = n2;
+    if (isUndefined(currParts)) {
+        return;
+    }
+
+    const { parts: prevParts } = n1;
+    if (currParts.length !== prevParts?.length) {
+        // It should not be possible to reach this point as patching only occurs when the vnode is the same.
+        if (process.env.NODE_ENV !== 'production') {
+            throw new Error('Expected static parts to be the same for the same element');
+        }
+        // Recover from patching issue
+        return applyStaticParts(root, n2, renderer, false);
+    }
```

Comment:
This should technically not be reachable because patch should only occur if a new vnode is generated that's the same as the previous one.

If the vnodes are reused we [skip the patching process](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L89-L91).

I'm adding this part more as a cautionary check in case something goes wrong and we have a way to recover from it.

I thought about just throwing an error here but it doesn't seem like it would be useful to a component author as there's probably nothing they can do to fix it.

Reply:
I understand the desire for the safeguard, but if we just `throw Error` without calling `applyStaticParts` (similar to [how we handle it](https://github.com/salesforce/lwc/blob/090aab8f59209249a677369965781da795578b7b/packages/%40lwc/engine-core/src/framework/modules/static-parts.ts#L61-L67) in `static-parts.ts`), then `applyStaticParts` will only ever be called with `mount = true`, and we can just rename it to `mountStaticParts` and simplify the code.

I think we can be reasonably certain that this code won't ever be hit, so I would say YAGNI and go with the simpler version for now. If not, we know how to fix in patch. 🙂

Reply:
@nolanlawson I thought about it and I think you're right we probably won't hit this safeguard. 

If we do run into this issue, it would be an issue with patching in general and more specifically why we are attempting to patch two different vnodes.

The error handling should be done higher in the call stack in that scenario.

I updated in the latest [commit](https://github.com/salesforce/lwc/pull/3998/commits/b73b67ba596311b903afb47c7795bf5ffe806965#diff-0f0f1360cc9d177fad1e943a03359f9e949ddfd86c8084e092c4811cbccf1484R119-L141)!